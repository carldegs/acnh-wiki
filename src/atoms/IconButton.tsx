import Link from 'next/link';
import { ReactElement } from 'react';

import { EmotionComponentProps } from '../types/EmotionComponentProps';

interface IconButtonProps extends EmotionComponentProps {
  onClick?: () => void;
  icon: ReactElement;
  href?: string;
  id?: string;
}

const IconButton: React.FC<IconButtonProps> = ({
  icon,
  href,
  onClick,
  className,
  id,
}) => {
  const Button = (
    <button
      css={(theme) => ({
        background: theme.colors.primary,
        border: 'none',
        borderRadius: '25% / 30%;',
        width: '100%',
        aspectRatio: '1',
        cursor: 'pointer',
      })}
      onClick={() => {
        if (onClick) {
          onClick();
        }
      }}
      className={className}
      data-testid={`button-${id}`}
    >
      {icon}
    </button>
  );

  return href ? <Link href={href}>{Button}</Link> : Button;
};

export default IconButton;
