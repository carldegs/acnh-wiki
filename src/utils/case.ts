export const kebabToCamel = (str: string) =>
  str.replace(/-./g, (match) => match[1].toUpperCase());

export const snakeToCamel = (str: string) =>
  str.replace(/_./g, (match) => match[1].toUpperCase());

export const jsonToCamel = <T = Record<string, any>>(obj: T): T =>
  Object.fromEntries(
    Object.entries(obj).map(([key, val]) => [
      snakeToCamel(kebabToCamel(key)),
      !Array.isArray(val) && typeof val === 'object' ? jsonToCamel(val) : val,
    ])
  ) as T;
