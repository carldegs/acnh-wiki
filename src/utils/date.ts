import { MONTHS } from '../constants/date';

export const getMonthRange = (monthRangeStr: string) => {
  const seasonsSplit = monthRangeStr.split(' & ').filter((s) => s !== '');

  if (!seasonsSplit.length) {
    return [];
  }

  return seasonsSplit.map((season) => {
    const periodSplit = season.split('-');

    return `${MONTHS[Number(periodSplit[0])]} - ${
      MONTHS[Number(periodSplit[1])]
    }`;
  });
};
