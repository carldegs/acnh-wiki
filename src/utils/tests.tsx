import { Global, ThemeProvider } from '@emotion/react';
import { render, RenderOptions } from '@testing-library/react';
import { rest } from 'msw';
import { ReactElement } from 'react';
import { QueryClientProvider, Hydrate, QueryClient } from 'react-query';

import { mockFishList } from '../mocks/fish';
import theme from '../theme';

export const handlers = [
  rest.get('*/fish', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockFishList));
  }),
  rest.get('*/fish/bitterling', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockFishList[0]));
  }),
];

export const ProvidersWrapper: React.FC = ({ children }) => {
  return (
    <ThemeWrapper>
      <QueryWrapper>{children}</QueryWrapper>
    </ThemeWrapper>
  );
};

export const ThemeWrapper: React.FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Global styles={theme.global} />
      {children}
    </ThemeProvider>
  );
};

export const QueryWrapper: React.FC = ({ children }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  return (
    <QueryClientProvider client={queryClient}>
      <Hydrate state={{}}>{children}</Hydrate>
    </QueryClientProvider>
  );
};

export const renderClientWithQuery = (ui: React.ReactElement) => {
  const testQueryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  const { rerender, ...result } = render(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>,
    { wrapper: ThemeWrapper }
  );

  return {
    ...result,
    rerender: (rerenderUi: React.ReactElement) =>
      rerender(
        <QueryClientProvider client={testQueryClient}>
          {rerenderUi}
        </QueryClientProvider>
      ),
  };
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: ProvidersWrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
