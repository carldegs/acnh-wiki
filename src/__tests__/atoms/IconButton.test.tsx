import IconButton from '../../atoms/IconButton';
import { render, fireEvent } from '../../utils/tests';

describe('IconButton', () => {
  it('renders items page unchanged', () => {
    const { container } = render(<IconButton icon={<div>+</div>} />);

    expect(container).toMatchSnapshot();
  });

  it('shows the icon', () => {
    const screen = render(<IconButton icon={<div>+</div>} />);

    expect(screen.getByText('+')).toBeInTheDocument();
  });

  it('triggers onClick function', () => {
    const func = jest.fn();
    const screen = render(<IconButton icon={<div>+</div>} onClick={func} />);

    fireEvent.click(screen.getByText('+'));

    expect(func).toHaveBeenCalled();
  });
});
