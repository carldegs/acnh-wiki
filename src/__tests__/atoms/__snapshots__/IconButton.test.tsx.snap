// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton renders items page unchanged 1`] = `
.emotion-0 {
  background: #756c5b;
  border: none;
  border-radius: 25%/30%;
  width: 100%;
  aspect-ratio: 1;
  cursor: pointer;
}

<div>
  <button
    class="emotion-0"
    data-testid="button-undefined"
  >
    <div>
      +
    </div>
  </button>
</div>
`;
